src/algebra/group_power/lemmas.lean

```diff
@@ -541,6 +541,34 @@ lemma le_self_sq (b : ℤ) : b ≤ b ^ 2 := le_trans (le_nat_abs) (abs_le_self_s
 
 alias int.le_self_sq ← int.le_self_pow_two
 
+lemma pow_right_injective {x : ℤ} (h : 2 ≤ x.nat_abs) : function.injective (λ (n : ℕ), x ^ n) :=
+λ n m hnm, begin
+  obtain pos | neg := nat_abs_eq x,
+  { lift x to ℕ using (pos.symm ▸ coe_nat_nonneg _),
+  norm_cast at h hnm,
+  exact nat.pow_right_injective h hnm },
+  dsimp only at hnm,
+  rw [←neg_neg x, neg_pow, neg_pow (-x)] at hnm,
+  replace neg := neg_eq_iff_neg_eq.mp neg.symm,
+  generalize' hy : -x = y,
+  rw [←neg_neg x, nat_abs_neg] at h,
+  rw [hy] at *,
+  lift y to ℕ using (neg.symm ▸ coe_nat_nonneg _),
+  obtain h₁ | h₁ := neg_one_pow_eq_or ℤ n;
+  obtain h₂ | h₂ := neg_one_pow_eq_or ℤ m;
+  all_goals { simp only [h₁, h₂, neg_mul_eq_neg_mul_symm, one_mul, neg_inj] at hnm,
+              norm_cast at h hnm,
+              exact nat.pow_right_injective h hnm <|> exfalso },
+  swap,
+  replace hnm := hnm.symm,
+  all_goals { apply int.ne_neg_of_pos _ _ hnm;
+              norm_cast;
+              apply nat.pos_of_ne_zero;
+              apply pow_ne_zero;
+              apply ne_bot_of_gt;
+              exact nat.succ_le_iff.mp h }
```

Comment:
Sounds good. This proof is a bit too complicated for me to understand in its entirety. I'm happy to add these changes.

Reply:
All I really changed was to replace your `;` with an `all_goals`, and to use `case` to be precise about which goals I was solving rather than your `<|>` to just keep going until they were all gone.